### PR TITLE
Dataservices api key

### PR DESCRIPTION
**Pull request by oleurud**
Solves #14263 

- [x] API Key implementation in Auth API
- [x] Save info in cdb-conf when appies
- [x] Remove info from cdb-conf when appies
- [x] Tests

### CODE REVIEW

File: app/models/carto/api_key.rb [hunk 1]
```diff
@@ -292,6 +321,13 @@ def process_table_permissions
       table_permissions
     end
 
+    def process_data_services
```
Comment (juanignaciosl) [inline app/models/carto/api_key.rb]: Minor: there's not a lot of processing here, `get_data_services` might be more accurate.

File: app/models/carto/api_key.rb [hunk 2]
```diff
@@ -1,12 +1,16 @@
...
 require_dependency 'carto/errors'
 require_dependency 'carto/helpers/auth_token_generator'
+require 'json'
 
 class ApiKeyGrantsValidator < ActiveModel::EachValidator
   def validate_each(record, attribute, value)
     return record.errors[attribute] = ['grants has to be an array'] unless value && value.is_a?(Array)
     record.errors[attribute] << 'only one apis section is allowed' unless value.count { |v| v[:type] == 'apis' } == 1
     record.errors[attribute] << 'only one database section is allowed' if value.count { |v| v[:type] == 'database' } > 1
+    if value.count { |v| v[:type] == 'dataservices' } > 1
```
Comment (juanignaciosl) [inline app/models/carto/api_key.rb]: Nice check. Must other types be unique as well? If that's the case, maybe you can generalize this to fail if any type is repeated. Otherwise it looks like dataservices is the only unique one.

File: app/models/carto/api_key.rb [hunk 3]
```diff
@@ -250,6 +264,21 @@ def set_enabled_for_engine
       valid_user? ? add_to_redis : remove_from_redis
     end
 
+    def save_dataservices_cdb_conf
+      CartoDB::Logger.debug(message: 'save_dataservices_cdb_conf')
+
+      info = {
+        "application": '',
```
Comment (juanignaciosl) [inline app/models/carto/api_key.rb]: :thinking: Empty application? If this API key is not related to an application, maybe the field shouldn't be there.

_opening Pandora's Box_: is dataservices permission configuration the right level for the application that an API key is related to? :thinking:

File: app/models/carto/api_key.rb [hunk 4]
```diff
@@ -250,6 +264,21 @@ def set_enabled_for_engine
       valid_user? ? add_to_redis : remove_from_redis
     end
 
+    def save_dataservices_cdb_conf
+      CartoDB::Logger.debug(message: 'save_dataservices_cdb_conf')
```
[the remark this answers is not in the document]
Comment (javitonino) [reply app/models/carto/api_key.rb]: This would be too noisy (aka: eat our Rollbar quota). I'd remove it.

File: app/models/carto/api_key.rb (hunk 3 above)
Comment (javitonino) [reply app/models/carto/api_key.rb]: First of all: YAGNI. We can add applications here later if it's actually needed. It's not yet needed so let's get rid of it. But it's nice to see how this could be extended to implement it.

Answering to Nacho: It's probably not great, but if we need DS to write this to a log, it'll probably end up here, since it seems the best place to pass things from Rails to DS. If you feel putting permissions here is something that could bite us when/if we have to add something else here (most likely application id), let's discuss it, since adding this to the metrics is probably something we are going to tackle in the next few months. Otherwise, let's discuss it again when it comes up.

Comment (juanignaciosl) [reply app/models/carto/api_key.rb]: No need to discuss it now, LDIAWICU :-)

File: app/controllers/carto/api/api_key_presenter.rb [hunk 5]
```diff
@@ -9,21 +9,31 @@ def initialize(api_key)
 
       def to_poro
         return {} unless @api_key
+
+        grants = [
```
[the remark this answers is not in the document]
Comment (oleurud) [reply app/controllers/carto/api/api_key_presenter.rb]: :+1:

File: app/models/carto/api_key.rb (hunk 4 above)
[the remark this answers is not in the document]
Comment (oleurud) [reply app/models/carto/api_key.rb]: Sorry, I forgot the remove it

File: app/models/carto/api_key.rb (hunk 3 above)
Comment (oleurud) [reply app/models/carto/api_key.rb]: Ok. I coded it before reading the comments on https://github.com/CartoDB/dataservices-api/pull/519

File: app/models/carto/api_key.rb (hunk 1 above)
Comment (oleurud) [reply app/models/carto/api_key.rb]: I named it following the similar methods: `process_table_permissions` and `process_granted_apis`

File: app/models/carto/api_key.rb [hunk 6]
```diff
@@ -292,6 +321,13 @@ def process_table_permissions
       table_permissions
     end
 
+    def process_data_services
+      data_services_grants = grants.find { |v| v[:type] == 'dataservices' }
```
[the remark this answers is not in the document]
Comment (oleurud) [reply app/models/carto/api_key.rb]: I not sure about this function, but again, I have tried to follow the same idea as `process_table_permissions` and `process_granted_apis`.

So, the `v` could be: `apis`, `database` or `dataservices` that are the valid parameters https://github.com/CartoDB/cartodb/pull/14270/files#diff-af4bff6ade71c9aa15290777d30c7fa4R13

The `apis` and `database` are required, `dataservices` doesn't

File: app/models/carto/api_key.rb (hunk 2 above)
Comment (oleurud) [reply app/models/carto/api_key.rb]: Initially, I made the same as `database` but the line was too long and I made this refactor. I will try to make in the same way as the other parameters

File: app/models/carto/api_key.rb [hunk 7]
```diff
@@ -250,6 +264,18 @@ def set_enabled_for_engine
       valid_user? ? add_to_redis : remove_from_redis
     end
 
+    def save_dataservices_cdb_conf
+      info = {
+        "permissions": data_services
```
Comment (javitonino) [inline app/models/carto/api_key.rb]: Shouldn't this include also the `username`? I left a similar comment at the DS side of things.

Also, more Ruby-ish way to write hashes: 
```ruby
info = {
  permissions: data_services
}
```

File: app/controllers/carto/api/api_key_presenter.rb [hunk 8]
```diff
@@ -9,28 +9,42 @@ def initialize(api_key)
...
+            type: 'apis',
+            apis: @api_key.granted_apis
+          },
+          {
+            type: 'database',
+            tables: table_permissions_for_api_key
+          }
+        ]
+
+        if @api_key.data_services.present?
```
Comment (javitonino) [inline app/controllers/carto/api/api_key_presenter.rb]: You have a helper method `data_services?` that is equivalent to this `present?` check. Maybe you can use it?

File: app/models/carto/api_key.rb [hunk 9]
```diff
@@ -1,12 +1,16 @@
 require 'securerandom'
 require_dependency 'carto/errors'
 require_dependency 'carto/helpers/auth_token_generator'
+require 'json'
```
Comment (javitonino) [inline app/models/carto/api_key.rb]: Just curious: is this needed?

File: app/models/carto/api_key.rb [hunk 10]
```diff
@@ -214,6 +224,10 @@ def needs_setup?
       regular? || oauth?
     end
 
+    def data_services?
+      self.data_services.present?
```
Comment (javitonino) [inline app/models/carto/api_key.rb]: `self` is redundant here :)

File: app/models/carto/api_key.rb [hunk 11]
```diff
@@ -250,6 +264,18 @@ def set_enabled_for_engine
       valid_user? ? add_to_redis : remove_from_redis
     end
 
+    def save_dataservices_cdb_conf
+      info = {
+        "permissions": data_services
+      }
+
+      db_run("SELECT cartodb.cdb_conf_setconf('#{db_role}', '#{info.to_json}');")
```
Comment (javitonino) [inline app/models/carto/api_key.rb]: In the [DS side of things](https://github.com/CartoDB/dataservices-api/pull/519/files#diff-cc3cf5f676070cffd04dcbae940b2059R22), the key is called `api_keys_{db_role}`. You should sync in the naming schema

cc @antoniocarlon

File: app/models/carto/api_key.rb [hunk 12]
```diff
@@ -292,6 +318,13 @@ def process_table_permissions
       table_permissions
     end
 
+    def process_data_services
+      data_services_grants = grants.find { |v| v[:type] == 'dataservices' }
+      return nil unless data_services_grants.present?
+
+      grants.find { |v| v[:type] == 'dataservices' }[:services]
```
Comment (javitonino) [inline app/models/carto/api_key.rb]: Try putting `grants.find { |v| v[:type] == 'dataservices' }` into a variable since it's calculated twice.

File: app/models/carto/api_key.rb (hunk 7 above)
Comment (oleurud) [reply app/models/carto/api_key.rb]: About the `username`, we discuss about it in
https://github.com/CartoDB/dataservices-api/issues/518#issuecomment-418732200 

As a resume, we have a way to extract the user (and organization). If we add a new way to extract it, we will need to add more logic in the data services functions in order to know if we should use the current way or the new way.

File: app/controllers/carto/api/api_key_presenter.rb (hunk 8 above)
Comment (oleurud) [reply app/controllers/carto/api/api_key_presenter.rb]: :+1:

File: app/models/carto/api_key.rb (hunk 11 above)
Comment (oleurud) [reply app/models/carto/api_key.rb]: :+1:

File: app/models/carto/api_key.rb (hunk 10 above)
Comment (oleurud) [reply app/models/carto/api_key.rb]: :+1:

File: app/models/carto/api_key.rb (hunk 12 above)
Comment (oleurud) [reply app/models/carto/api_key.rb]: :+1:

File: app/models/carto/api_key.rb (hunk 9 above)
Comment (oleurud) [reply app/models/carto/api_key.rb]: I thought so. But after your comment, I have checked that it is not. Thanks

**Comment by javitonino** [comment]
Cannot do CR while tests are broken (since there will be changes to fix them) :upside_down_face:

**Comment by javitonino** [comment]
Task running...
